### PR TITLE
[feature](functions) impl scalar functions trim_in、ltrim_in and rtrim_in

### PR DESCRIPTION
**Pull request by liujiwen-up**
trim_in is different from trim
Find and remove any characters in a set of characters at both ends of a string (regardless of order)

mysql> SELECT TRIM('abcd', 'cde');
+---------------------+
| trim('abcd', 'cde') |
+---------------------+
| abcd                |
+---------------------+
1 row in set (0.02 sec)

mysql> SELECT TRIM_IN('abcd', 'cde');
+------------------------+
| trim_in('abcd', 'cde') |
+------------------------+
| ab                     |
+------------------------+
1 row in set (0.02 sec)

### CODE REVIEW

**Comment by liujiwen-up** [comment]
run buildall

File: regression-test/suites/query_p0/sql_functions/string_functions/test_trim_in.groovy [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    order_qt_46 "SELECT ltrim_in(CAST(' hello world ' AS CHAR(13)), ' ehlowrd');"
+    order_qt_47 "SELECT ltrim_in(CAST(' hello world ' AS CHAR(13)), ' x');"
+    order_qt_48 "SELECT rtrim_in('', '');"
+    order_qt_49 "SELECT rtrim_in('   ', '');"
+    order_qt_50 "SELECT rtrim_in('  hello  ', '');"
+    order_qt_51 "SELECT rtrim_in('  hello  ', ' ');"
+    order_qt_52 "SELECT rtrim_in('  hello  ', 'lo ');"
+    order_qt_53 "SELECT rtrim_in('hello  ', ' ');"
+    order_qt_54 "SELECT rtrim_in('hello  ', 'l o');"
+    order_qt_55 "SELECT rtrim_in('hello  ', 'l');"
```
Comment (yiguolei) [inline regression-test/suites/query_p0/sql_functions/string_functions/test_trim_in.groovy:135]: add more test
1. create a table and then insert the values to a table, and use select func(col) from table
2. the column in the table, could be a nullable column and not null column.

File: be/src/vec/functions/function_string.cpp [hunk 2]
```diff
@@ -535,6 +545,42 @@ struct TrimUtil {
         return Status::OK();
     }
 };
+template <bool is_ltrim_in, bool is_rtrim_in, bool trim_single>
+struct TrimInUtil {
+    static Status vector(const ColumnString::Chars& str_data,
```
Comment (yiguolei) [inline be/src/vec/functions/function_string.cpp:550]: 为什么不直接使用column string，而是把chars和offsets 分开

Comment (liujiwen-up) [reply be/src/vec/functions/function_string.cpp:550]: In order to maintain the same processing method as the TrimUtil method, so no changes are made here.
`struct TrimUtil {
    static Status vector(const ColumnString::Chars& str_data,
                         const ColumnString::Offsets& str_offsets, const StringRef& remove_str,
                         ColumnString::Chars& res_data, ColumnString::Offsets& res_offsets)`

File: regression-test/suites/query_p0/sql_functions/string_functions/test_trim_in.groovy (hunk 1 above)
Comment (liujiwen-up) [reply regression-test/suites/query_p0/sql_functions/string_functions/test_trim_in.groovy:164]: done

File: be/src/vec/functions/function_string.cpp [hunk 3]
```diff
@@ -535,6 +545,42 @@ struct TrimUtil {
...
+                         const ColumnString::Offsets& str_offsets, const StringRef& remove_str,
+                         ColumnString::Chars& res_data, ColumnString::Offsets& res_offsets) {
+        const size_t offset_size = str_offsets.size();
+        res_offsets.resize(offset_size);
+        res_data.reserve(str_data.size());
+        std::bitset<256> char_lookup;
```
Comment (zhiqiang-hhhh) [inline be/src/vec/functions/function_string.cpp:556]: Add comment to explain magic number 256

File: be/src/vec/functions/function_string.cpp [hunk 4]
```diff
@@ -535,6 +545,42 @@ struct TrimUtil {
...
+        for (size_t i = 0; i < remove_str.size; ++i) {
+            char_lookup[static_cast<unsigned char>(remove_str.data[i])] = true;
```
Comment (zhiqiang-hhhh) [inline be/src/vec/functions/function_string.cpp:558]: what if chinese

File: be/src/vec/functions/function_string.cpp (hunk 3 above)
Comment (liujiwen-up) [reply be/src/vec/functions/function_string.cpp]: Currently, there are two processing logics. 
1. The `simd::VStringFunctions::is_ascii `method is used to judge that when the string is all ascii, bitset<128> will be used for processing. The character range of the standard ASCII table is 0 to 127. Bitset<128> has exactly 128 bits, which is enough to represent all standard ASCII characters. 
2. When the string is not all standard ascii, the utf-8 logic will be used for processing.It is especially important to note that when trimming on the right, according to the rules of UTF-8, the format of the UTF-8 trailing byte is 10xxxxxx. Use `(*prev_char_pos & 0xC0) == 0x80` to determine whether the current byte is a trailing byte. When a byte that is not a trailing byte is found, this byte is the starting byte of the current character.

File: be/src/vec/functions/function_string.cpp (hunk 4 above)
Comment (liujiwen-up) [reply be/src/vec/functions/function_string.cpp]: Added utf-8 processing logic

**Review by zhiqiang-hhhh — APPROVED** [review]
LGTM

**Comment by zhiqiang-hhhh** [comment]
run buildall

File: be/src/vec/functions/function_string.cpp [hunk 5]
```diff
@@ -535,6 +545,135 @@ struct TrimUtil {
...
+        const char* remove_end = remove_str.data + remove_str.size;
+
+        while (remove_begin < remove_end) {
+            char_lookup.set(static_cast<unsigned char>(*remove_begin));
+            remove_begin += 1;
+        }
+
+        for (size_t i = 0; i < offset_size; ++i) {
+            const char* str_begin = reinterpret_cast<const char*>(
+                    str_data.data() + (i == 0 ? 0 : str_offsets[i - 1]));
```
[the remark this answers is not in the document]
Comment (liujiwen-up) [reply be/src/vec/functions/function_string.cpp]: done

File: be/src/vec/functions/function_string.cpp [hunk 6]
```diff
@@ -535,6 +545,135 @@ struct TrimUtil {
...
+
+    static Status impl_vectors_utf8(const ColumnString::Chars& str_data,
+                                    const ColumnString::Offsets& str_offsets,
+                                    const StringRef& remove_str, ColumnString::Chars& res_data,
+                                    ColumnString::Offsets& res_offsets) {
+        const size_t offset_size = str_offsets.size();
+        res_offsets.resize(offset_size);
+        res_data.reserve(str_data.size());
+
+        std::unordered_set<std::string> char_lookup;
```
[the remark this answers is not in the document]
Comment (liujiwen-up) [reply be/src/vec/functions/function_string.cpp]: Change to string_view

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by HappenLee** [comment]
Please add doc in https://github.com/apache/doris-website @liujiwen-up

File: be/src/vec/functions/function_string.cpp [hunk 7]
```diff
@@ -640,6 +788,38 @@ class FunctionTrim : public IFunction {
     }
 };
 
+template <typename impl>
```
Comment (HappenLee) [inline be/src/vec/functions/function_string.cpp:791]: Impl

File: be/src/vec/functions/function_string.cpp [hunk 8]
```diff
@@ -583,14 +722,23 @@ struct Trim2Impl {
...
+
                 if (remove_str.size == 1) {
                     RETURN_IF_ERROR((TrimUtil<is_ltrim, is_rtrim, true>::vector(
                             col->get_chars(), col->get_offsets(), remove_str, col_res->get_chars(),
                             col_res->get_offsets())));
                 } else {
-                    RETURN_IF_ERROR((TrimUtil<is_ltrim, is_rtrim, false>::vector(
-                            col->get_chars(), col->get_offsets(), remove_str, col_res->get_chars(),
-                            col_res->get_offsets())));
+                    if constexpr (std::is_same<Name, NameTrimIn>::value ||
```
Comment (HappenLee) [inline be/src/vec/functions/function_string.cpp:731]: `TrimInUtil` and `TrimUtil` better be a template param

File: be/src/vec/functions/function_string.cpp (hunk 7 above)
Comment (liujiwen-up) [reply be/src/vec/functions/function_string.cpp:791]: done

File: be/src/vec/functions/function_string.cpp (hunk 8 above)
Comment (liujiwen-up) [reply be/src/vec/functions/function_string.cpp:731]: done

**Review by HappenLee — APPROVED** [review]
LGTM

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by liujiwen-up** [comment]
run buildall

**Comment by liujiwen-up** [comment]
> Please add doc in https://github.com/apache/doris-website @liujiwen-up

https://github.com/apache/doris-website/pull/1207

**Comment by liujiwen-up** [comment]
run external

**Pull request merged**